packages/jest-haste-map/src/crawlers/watchman.js

```diff
@@ -18,126 +18,121 @@ import H from '../constants';
 const watchmanURL =
   'https://facebook.github.io/watchman/docs/troubleshooting.html';
 
-function isDescendant(root: string, child: string): boolean {
-  return child.startsWith(root);
-}
-
 function WatchmanError(error: Error): Error {
-  return new Error(
+  error.message =
     `Watchman error: ${error.message.trim()}. Make sure watchman ` +
-      `is running for this project. See ${watchmanURL}.`,
-  );
+    `is running for this project. See ${watchmanURL}.`;
+  return error;
 }
 
-module.exports = function watchmanCrawl(
+module.exports = async function watchmanCrawl(
   options: CrawlerOptions,
 ): Promise<InternalHasteMap> {
   const {data, extensions, ignore, roots} = options;
-  // Watchman always returns POSIX style paths so use posixRoots
-  // instead of roots to avoid on-the-fly checks inside the loop.
-  const posixRoots =
-    path.sep === '/'
-      ? Array.from(roots)
-      : roots.map(root => root.replace(/\\/g, '/'));
+  const defaultWatchExpression = [
+    'allof',
+    ['type', 'f'],
+    ['anyof'].concat(extensions.map(extension => ['suffix', extension])),
+  ];
+  const client = new watchman.Client();
+  let clientError;
+  client.on('error', error => (clientError = error));
 
-  return new Promise((resolve, reject) => {
-    const client = new watchman.Client();
-    client.on('error', error => reject(error));
+  const cmd = (...args) =>
+    new Promise((resolve, reject) =>
+      client.command(
+        args,
+        (error, result) => (error ? reject(error) : resolve(result)),
+      ),
+    );
 
-    const cmd = (...args) =>
-      new Promise((resolve, reject) => {
-        client.command(args, (error, result) => {
-          if (error) {
-            reject(error);
-          } else {
-            resolve(result);
-          }
-        });
-      });
+  const clocks = data.clocks;
+  let files = data.files;
 
-    const clocks = data.clocks;
-    let files = data.files;
+  try {
```

Comment:
this is a gigantic try/catch, and really hard to read. Was it infeasible to break it up into functions?

Reply:
Well, it is because I tried to mimic the earlier behavior which was not very visible. This is also mostly to wrap the error in that custom text we have. I can try narrowing this down at the risk not not being fully backwards compatible regarding error handling and error message customization.